workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt

```diff
@@ -197,18 +203,33 @@ inline fun <StateT, OutputT : Any> action(
  * @see StatelessWorkflow.action
  * @see StatefulWorkflow.action
  */
-inline fun <StateT, OutputT : Any> action(
+inline fun <StateT, OutputT> action(
   crossinline name: () -> String,
   crossinline apply: Updater<StateT, OutputT>.() -> Unit
 ): WorkflowAction<StateT, OutputT> = object : WorkflowAction<StateT, OutputT> {
   override fun Updater<StateT, OutputT>.apply() = apply.invoke(this)
   override fun toString(): String = "WorkflowAction(${name()})@${hashCode()}"
 }
 
-fun <StateT, OutputT : Any> WorkflowAction<StateT, OutputT>.applyTo(
+// TODO Is this new "pass me a lambda" API super gross?
+// @OptIn(ExperimentalContracts::class)
+// inline fun <T, StateT, OutputT> WorkflowAction<StateT, OutputT>.applyTo(
+//   state: StateT,
+//   block: (
+//     newState: StateT,
+//     hasOutput: Boolean,
+//     output: OutputT
+//   ) -> T
+// ): T {
+//   contract { callsInPlace(block, EXACTLY_ONCE) }
+//   val updater = Updater<StateT, OutputT>(state)
+//   updater.apply()
+//   return block(updater.nextState, updater._output !== NoOutput, updater._output)
+// }
+fun <StateT, OutputT> WorkflowAction<StateT, OutputT>.applyTo(
   state: StateT
-): Pair<StateT, OutputT?> {
+): Triple<StateT, Boolean, OutputT> {
   val updater = Updater<StateT, OutputT>(state)
   updater.apply()
-  return Pair(updater.nextState, updater.output)
+  return Triple(updater.nextState, updater.output !== NoOutput, updater.output)
```

Comment:
I kind of hate both these solutions. The `Triple` is getting confusing, the function feels like overengineering, and I really don't want to introduce yet another type just for the _result_ of an action. Any better ideas?

Reply:
Custom result type? At least it would be a bit more self documenting.

```
class WorkflowTransition<StateT, OutputT>(
  val newState: StateT,
  val output: OutputT,
  val hasOutput
)
```

Sealed class?

```
sealed class WorkflowTransition<StateT, OutputT> {
  class WithOutput(
    val newState: StateT,
    val output: OutputT
  ) : WorkflowTransition()

  class WithoutOutput(val newState: StateT) : WorkflowTransition
}
```

Create our own output-specific Maybe type?

```
sealed class OutputMaybe<OutputT> { ... }

class WorkflowTransition<StateT, OutputT>(
  val newState: StateT,
  val output: OutputMaybe<OutputT>
)
```

Reply:
That last one might be able to replace `NoOutput` too.

Reply:
> I really don't want to introduce yet another type just for the result of an action.

This type is gonna be part of our public API too.

Reply:
Is this a sign that nullable `OutputT` isn't worth the bother? It's been useful for `RenderT`, but that seems like a very different thing. `OutputT` is basically an event type, and a null event seems like a very smelly notion to me. Do I really want to be able to / forced to choose between emitting nothing and emitting null?

Reply:
Oh, you called this out as a GUWT blocker. :/ 

Hmm.

Reply:
Yea, it would be really unfortunate to regress back to RxJava's null limitations for no apparent reason (the reason being purely internal). 

> Do I really want to be able to / forced to choose between emitting nothing

Emitting null can definitely be useful for workers/streams, even if not for workflows themselves. And this restriction is entirely arbitrary from a consumer point of view, so it's kind of awkward.